src/Microsoft.DotNet.ProjectJsonMigration/ProjectDependencyFinder.cs

```diff
@@ -0,0 +1,204 @@
+                    {
+                        var projectExists = File.Exists(project.ProjectFilePath);
+                        var otherExists = File.Exists(otherProject.ProjectFilePath);
+
+                        if (projectExists != otherExists
+                            && projectExists
+                            && !otherExists)
+                        {
+                            // the project currently in the cache does not exist, but this one does
+                            // remove the old project and add the current one
+                            projects[project.Name] = project;
+                        }
+                    }
+                }
+                else
+                {
+                    projects.Add(project.Name, project);
+                }
+            }
+
+            return projects;
+        }
+
+        /// <summary>
+        /// Create the list of potential projects from the search paths.
+        /// </summary>
+        private static List<ProjectDependency> GetPotentialProjects(IEnumerable<string> searchPaths)
+        {
+            var projects = new List<ProjectDependency>();
+
+            // Resolve all of the potential projects
+            foreach (var searchPath in searchPaths)
+            {
+                var directory = new DirectoryInfo(searchPath);
+
+                if (!directory.Exists)
+                {
+                    continue;
+                }
+
+                foreach (var projectDirectory in directory.EnumerateDirectories())
+                {
+                    // Create the path to the project.json file.
+                    var projectFilePath = Path.Combine(projectDirectory.FullName, "project.json");
+
+                    // We INTENTIONALLY do not do an exists check here because it requires disk I/O
+                    // Instead, we'll do an exists check when we try to resolve
+
+                    // Check if we've already added this, just in case it was pre-loaded into the cache
+                    var project = new ProjectDependency(projectDirectory.Name, projectFilePath);
+
+                    projects.Add(project);
+                }
+            }
+
+            return projects;
+        }
+
+        private static List<string> GetGlobalPaths(string rootPath)
+        {
+            var paths = new List<string>();
+
+            var globalJsonRoot = ResolveRootDirectory(rootPath);
+
+            GlobalSettings globalSettings;
+            if (GlobalSettings.TryGetGlobalSettings(globalJsonRoot, out globalSettings))
+            {
+                foreach (var sourcePath in globalSettings.ProjectPaths)
+                {
+                    var path = Path.GetFullPath(Path.Combine(globalJsonRoot, sourcePath));
+
+                    paths.Add(path);
+                }
+            }
+
+            return paths;
+        }
+
+        public static string ResolveRootDirectory(string projectPath)
```

Comment:
To me, I think it would make more sense if the GlobalSettings class was responsible for walking up to the parent directories.  Someone outside the class could call `GlobalSettings.TryGetGlobalSettings(path, out...)`, and GlobalSettings did all the logic itself.

---
In reply to: [80042789](https://github.com/dotnet/cli/pull/4224#discussion_r80042789) [](ancestors = 80042789)